### PR TITLE
executor: make outer table can build hash table for hash join

### LINKED ISSUE #6868: use the smaller table to build hash table for hash join
## 1. What did you do?

For **left outer join** or **right outer join**, we chose the inner table to build the hash table for hash join currently. In fact, we can chose a smaller one to build the hash table even if it is the outer table.

Take this as an example:

```sql
drop table if exists t1, t2;
create table t1(a bigint, b bigint); -- the smaller one
create table t2(a bigint, b bigint); -- the bigger  one
select * from t1 left outer join t2 on t1.a = t2.a;
```

For now we chose `t2` to build the hash table even if it may contain more records than `t1`, which can somehow possibly cause the following problems:
1. more probably to run out of memory
2. worse hash conflict, and the worse the conflict is, the more probe time may take.

If we chose the smaller table `t1` to build hash table, we can avoid these problem.

## 2. What did you expect to see?

use the smaller table to build hash table and do hash join

## 3. What did you see instead?

```sql
TiDB(localhost:4000) > desc select * from t1 left outer join t2 on t1.a = t2.a;
+----------------+----------------+------------------------------+------+-------------------------------------------------------------------------+----------+
| id             | parents        | children                     | task | operator info                                                           | count    |
+----------------+----------------+------------------------------+------+-------------------------------------------------------------------------+----------+
| TableScan_7    |                |                              | cop  | table:t1, range:[-inf,+inf], keep order:false                           | 10000.00 |
| TableReader_8  | HashLeftJoin_6 |                              | root | data:TableScan_7                                                        | 10000.00 |
| TableScan_9    |                |                              | cop  | table:t2, range:[-inf,+inf], keep order:false                           | 10000.00 |
| TableReader_10 | HashLeftJoin_6 |                              | root | data:TableScan_9                                                        | 10000.00 |
| HashLeftJoin_6 |                | TableReader_8,TableReader_10 | root | left outer join, inner:TableReader_10, equal:[eq(test.t1.a, test.t2.a)] | 12500.00 |
+----------------+----------------+------------------------------+------+-------------------------------------------------------------------------+----------+
5 rows in set (0.00 sec)
```

## 4. What version of TiDB are you using?

```sh
➜ ./bin/tidb-server -V
Release Version: v2.1.0-alpha-55-g5022b5b2
Git Commit Hash: 5022b5b2cc31ef64b4115482aaf44a50a6618024
Git Branch: master
UTC Build Time: 2018-06-20 10:20:04
GoVersion: go version go1.10 darwin/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-c4133d3ef0f099f8716a916c198390b26293ec00
```

-----

## Seperate the problem into two parts:
- [ ] implement the execution part #12882
- [ ] generate the plan #12883

### PR DESCRIPTION
**Pull request by fzhedu**
### What problem does this PR solve? 
Fix #6868

### What is changed and how it works?
Reverse the inner and the outer for all outer joins when building hash joins if the outer size is smaller.
This function improves the left join by 30%, and the whole query by 20%  for TPC-H Q 13.
### Check List 

Tests 

 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by zz-jason** [comment]
@fzhedu please follow the [Commit Message and Pull Request Style](https://github.com/pingcap/community/blob/master/commit-message-pr-style.md) guide to reformat the PR title.

BTW, it's recommended to use a `fix` or `close` keyword to close issue via this PR, see [Closing issues using keywords](https://help.github.com/en/articles/closing-issues-using-keywords) for more detail.

File: executor/benchmark_test.go [hunk 1]
```diff
@@ -574,7 +574,97 @@ func prepare4Join(testCase *hashJoinTestCase, innerExec, outerExec Executor) *Ha
 	}
 	return e
 }
+func prepare4OuterJoin(testCase *hashJoinTestCase, innerExec, outerExec Executor) *HashJoinExec {
+	// reverse
+	var temp = innerExec
```
Comment (XuHuaiyu) [inline executor/benchmark_test.go]: We can use  
``` go
innerExec, outerExec = outerExec, innerExec
```

File: executor/builder.go [hunk 2]
```diff
@@ -998,6 +998,14 @@ func (b *executorBuilder) buildHashJoin(v *plannercore.PhysicalHashJoin) Executo
 		joinType:      v.JoinType,
 		isOuterJoin:   v.JoinType.IsOuterJoin(),
 		innerEstCount: v.Children()[v.InnerChildIdx].StatsCount(),
+		outerHashJoin: v.OuterHashJoin,
+	}
+	// reverse the inner and the outer
+	if e.outerHashJoin {
```
Comment (XuHuaiyu) [inline executor/builder.go]: Why do we need to reverse the outer and inner side?
We can use the outer child to build the hash table directly if outerHashJoin is true without any modification on the metadata of HashJoinExec.

File: executor/builder.go [hunk 3]
```diff
@@ -1025,13 +1030,17 @@ func (b *executorBuilder) buildHashJoin(v *plannercore.PhysicalHashJoin) Executo
 		e.outerFilter = v.LeftConditions
 		e.innerKeys = v.RightJoinKeys
 		e.outerKeys = v.LeftJoinKeys
-		if defaultValues == nil {
+	}
+	if defaultValues == nil {
+		if e.outerHashJoin {
```
Comment (XuHuaiyu) [inline executor/builder.go]: This is not correct here.
defaultValues is used to fill the null-vals for outerJoin.
The count of defaultValues is always equal to the `e.innerExec.Schema().Len()`

File: executor/benchmark_test.go (hunk 1 above)
Comment (fzhedu) [reply executor/benchmark_test.go]: adopted.

File: executor/builder.go (hunk 3 above)
Comment (fzhedu) [reply executor/builder.go]: When defaultValues == nil, it should be initialized to be null-vals according to the coorresponding  inner side. This follows the logic of the former codes. Here seems ok.

Comment (fzhedu) [reply executor/builder.go]: > This is not correct here.
> defaultValues is used to fill the null-vals for outerJoin.
> The count of defaultValues is always equal to the `e.innerExec.Schema().Len()`

When defaultValues == nil, it should be initialized to be null-vals according to the coorresponding inner side. This follows the logic of the former codes. Here seems ok.

File: executor/builder.go (hunk 2 above)
Comment (fzhedu) [reply executor/builder.go]: > Why do we need to reverse the outer and inner side?
> We can use the outer child to build the hash table directly if outerHashJoin is true without any modification on the metadata of HashJoinExec.

If following your way, the build and the probe have to change a lot of codes. Here, I just reverse the outer and the inner within the join, the output of the join remains the same. The previous codes can be reused as many as possible.

**Pull request closed without merging**